keps/sig-node/xxxx-container-oom-kill-mode/README.md

```diff
@@ -0,0 +1,729 @@
+- **Existing behavior**: Containers without this field specified will continue to use the kubelet's node-level configuration until v1.38, ensuring zero impact on existing workloads during migration.
+
+### Risks and Mitigations
+
+**Risk**: Inconsistent behavior between cgroup v1 and v2 environments.
+
+- **Mitigation**: Clear documentation about platform limitations, validation warnings when `Group` is used on cgroup v1.
+
+**Risk**: User confusion about the interaction between container and kubelet settings.
+
+- **Mitigation**: Comprehensive documentation with examples, clear precedence rules, kubectl describe showing effective OOM mode.
+
+**Risk**: Performance impact from additional cgroup configuration.
+
+- **Mitigation**: The setting is applied once at container creation, no runtime overhead.
+
+**Risk**: Breaking change for users relying on `singleProcessOOMKill` flag.
+
+- **Mitigation**: 3-release deprecation period (v1.35-v1.37), clear migration guide, warning logs in v1.36+, container-level config available before removal.
+
+## Design Details
+
+### API Changes
+
+#### Container API
+
+The changes will be made to the Container struct in `staging/src/k8s.io/api/core/v1/types.go`:
+
+```go
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+    // ... existing fields ...
+
+    // OOMKillMode specifies how the OOM killer behaves for this container.
+    // +featureGate=ContainerOOMKillMode
+    // +optional
+    OOMKillMode *OOMKillMode `json:"oomKillMode,omitempty" protobuf:"bytes,26,opt,name=oomKillMode"`
+}
+```
+
+The same field applies to `EphemeralContainer` for consistency.
+
+### Kubelet Implementation
+
+The kubelet will be modified in `pkg/kubelet/kuberuntime/kuberuntime_container_linux.go` to respect the container-level setting:
+
+```go
+func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(
+    ctx context.Context,
+    pod *v1.Pod,
+    container *v1.Container,
+    enforceMemoryQoS bool,
+) *runtimeapi.LinuxContainerResources {
+    // ... existing code ...
+
+    // v1.36+: Warn about deprecated flag
+    if utilfeature.DefaultFeatureGate.Enabled(features.ContainerOOMKillMode) &&
+       m.singleProcessOOMKill != nil {
+        klog.Warning("Flag --single-process-oom-kill is deprecated and will be removed in v1.38. " +
+                    "Please migrate to pod.spec.containers.oomKillMode field.")
+    }
+
+    // Determine OOM kill behavior
+    var useGroupKill bool
+
+    if container.OOMKillMode != nil {
+        // Container-level setting takes precedence
+        switch *container.OOMKillMode {
+        case v1.OOMKillModeSingle:
+            useGroupKill = false
+        case v1.OOMKillModeGroup:
+            if !isCgroup2UnifiedMode() {
+                klog.Warningf("Container %s/%s requests Group mode but system uses cgroup v1, falling back to single process",
```

Comment:
End users probably don’t look at kubelet logs much, so it would be nice to notify them via Kubernetes Events. However, with the current kubelet implementation, that might be tricky.